face/PortraitProcessing/java/PortraitProcessing.java

```diff
@@ -0,0 +1,186 @@
+    // Margin ratio on face crop
+    private static final float TOP_MARGIN_MAX = 0.75F;
+    private static final float BOTTOM_MARGIN_MAX = 0.75F;
+    private static final float LEFT_MARGIN_MAX = 1.5F;
+    private static final float RIGHT_MARGIN_MAX = 1.5F;
+
+    public static void main(String[] args) throws Exception {
+        System.out.println("Sample code for portrait processing with Azure Face API and Background Removal API.");
+
+        // Read image from URL
+        BufferedImage image = ImageIO.read(new URL(IMAGE_URL));
+
+        // Resize image and compress with JPEG to reduce overall latency.
+        int resizedWidth = image.getWidth();
+        int resizedHeight = image.getHeight();
+        if (resizedWidth > MAX_IMAGE_SIZE || resizedHeight > MAX_IMAGE_SIZE) {
+            if (resizedWidth > resizedHeight) {
+                resizedWidth = Math.min(resizedWidth, MAX_IMAGE_SIZE);
+                resizedHeight = (int) ((float) resizedHeight * ((float) resizedWidth / (float) image.getWidth()));
+            } else {
+                resizedHeight = Math.min(resizedHeight, MAX_IMAGE_SIZE);
+                resizedWidth = (int) ((float) resizedWidth * ((float) resizedHeight / (float) image.getHeight()));
+            }
+        }
+        BufferedImage bufferedImage = new BufferedImage(resizedWidth, resizedHeight, BufferedImage.TYPE_INT_RGB);
+        Graphics graphics = bufferedImage.getGraphics();
+        graphics.drawImage(image.getScaledInstance(resizedWidth, resizedHeight, Image.SCALE_SMOOTH), 0, 0, null);
+        graphics.dispose();
+        ByteArrayOutputStream imageMemoryStream = new ByteArrayOutputStream();
+        ImageWriter imageMemoryStreamWriter = ImageIO.getImageWritersByFormatName("jpg").next();
+        ImageWriteParam imageMemoryStreamParam = imageMemoryStreamWriter.getDefaultWriteParam();
+        imageMemoryStreamParam.setCompressionMode(ImageWriteParam.MODE_EXPLICIT);
+        imageMemoryStreamParam.setCompressionQuality(JPEG_QUALITY);
+        imageMemoryStreamWriter.setOutput(ImageIO.createImageOutputStream(imageMemoryStream));
+        imageMemoryStreamWriter.write(null, new IIOImage(bufferedImage, null, null), imageMemoryStreamParam);
+        imageMemoryStreamWriter.dispose();
+
+        // Detect faces in the image
+        FaceClient faceClient = new FaceClientBuilder().endpoint(FACE_ENDPOINT).credential(new KeyCredential(FACE_KEY)).buildClient();
+        DetectOptions options = new DetectOptions(DETECTION_MODEL, RECO_MODEL, false).setReturnFaceAttributes(FACE_ATTRIBUTES).setReturnFaceLandmarks(true);
+        List<FaceDetectionResult> detectedFaces = faceClient.detect(BinaryData.fromBytes(imageMemoryStream.toByteArray()), options);
```

Comment:
BinaryData support constructing fromStream directly, would it support current scenario?
https://learn.microsoft.com/en-us/java/api/com.azure.core.util.binarydata?view=azure-java-stable#com-azure-core-util-binarydata-fromstream(java-io-inputstream)

Reply:
What we get from ImageWriter is a ByteArray**Output**Stream, but what BinaryData.fromStream() expects is a ByteArray**Input**Stream. Conversion between the two basically needs the extraction of underlaying byte array using method .toByteArray(), so it will look like this:

```diff
- List<FaceDetectionResult> detectedFaces = faceClient.detect(BinaryData.fromBytes(imageMemoryStream.toByteArray()), options);
+ InputStream stream = new ByteArrayInputStream(imageMemoryStream.toByteArray());
+ List<FaceDetectionResult> detectedFaces = faceClient.detect(BinaryData.fromStream(stream), options);
```

As you can see, it's a bit more complicated than the current one.